fix(compiler): no error when trying to reassign to a `let` variable

yoav-steinberg opened the pull request:
now this:
```wing
let x = 9;
x = 10;
```
produces something like: `Error at ../../examples/tests/invalid/reassing_to_nonreassignable.w:2:1 | Variable x is not reassignable `

Also correctly import reassignability (let/var) from JSII class properties.

see #1186


*By submitting this pull request, I confirm that my contribution is made under the terms of the 
[Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

yoav-steinberg commented:
Note that this us currently failing because one of our tests attempt to write to an argument passed to a function, but there's currently no way of specifying if an argument is reassignable:
```wing
let f = (x: num) => {
  x = x +1; // This write fails because `x` is non reassignable
};
```
Should we suppor the following syntax to fix this? @eladb @Chriscbr  WDYT?
```wing
let f = (var x: num) => {
  x = x +1;
};
```

Chriscbr commented:
@yoav-steinberg I'd be okay adding that syntax later if there is demand for it, though I feel like it's generally safer for parameters to not be reassignable by default.

Chriscbr commented on libs/wingc/src/type_check.rs:
```diff
@@ -1458,7 +1500,10 @@ impl<'a> TypeChecker<'a> {
 						span: name.span.clone(),
 					},
-					SymbolKind::Variable(constructor_type, VariableKind::Let),
+					SymbolKind::Variable(VariableInfo {
+						_type: constructor_type,
+						kind: VariableKind::Let,
```
Add a helper perhaps?

SymbolKind::make_variable(constructor_type, true)

Chriscbr reviewed (approved):
One suggestion to reduce some verbosity - otherwise LGTM. I like the var argument idea as well 👍

yoav-steinberg commented:
Update: fixed everything related to to function arguments not being `var`. 
Also added support for modifying non-var class fields in `init` so we can initialize classes (this also caused test failures).
Also added support (and tests) for checking reassignablity of class fields.
Additionally added support for mutable JSII class fields by marking them as reassignable on import.

Chriscbr commented on examples/tests/valid/reassignment.w:
```diff
@@ -3,3 +3,25 @@ assert(x == 5);
+  arg = 0;
+  return arg;
+}
+
+assert(f(1) == 0);
```
nit: It could be worth testing this case:
```suggestion

let y = 1;
assert(f(y) == 0);
assert(y == 1); // y is not modified, since Wing functions are pass-by-value

```

yoav-steinberg replied:
Why not, i'll add it.

eladb commented:
@yoav-steinberg is this ready to be merged?

yoav-steinberg commented:
> @yoav-steinberg is this ready to be merged?

Not yet, I woke up this morning having a feeling this might break:
```wing
resource R {
  f: num;
  init() {
    if true {
      this.f = 1; // Access a non reassignable field from constructor but not in upper most scope. Need to make sure this works!
    }
  }
}
